### PR TITLE
Scale on OOMs for autoscaling components

### PR DESCRIPTION
**Pull request by chencs**
#### What this PR does
This PR updates the memory scaling trigger for autoscaling components so that they treat OOMing containers as utilizing their entire memory request. It also adds a memory scaling trigger to ruler-queriers, which previously only scaled on CPU. Previously, if ruler-queriers OOMed and CPU usage went down, they would scale down, putting more memory pressure on the remaining ruler-queriers. 

This also updates the remote ruler reads dashboard to fix a panel query bug, and separates panels for ruler-querier CPU and memory scaling metrics.

#### Which issue(s) this PR fixes or relates to
(none)

#### Checklist

- [x] Tests updated
`N/A` Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by jhesketh** [comment]
This vaguely looks correct to me in terms of adding a new trigger. It's the query that'll be the hard part to figure out.

I agree that having a flag to switch OOM scaling on/off is a good idea. We'd likely then want to do it for most (all?) of our HPA's, at a guess.

**Comment by chencs** [comment]
At this point, `newRulerQuerierScaledObject` and `newQueryFrontendScaledObject` are similar enough that I wonder why they shouldn't be condensed into one function, where things that used to use `newQueryFrontendScaledObject` simply call the new function with an explicit `weight` of 1.

**Comment by chencs** [comment]
I can add another flag to optionally add the ruler-querier memory scaling trigger just for rolling changes out, wasn't sure if we'd need that.

File: operations/mimir/autoscaling.libsonnet [hunk 1]
```diff
@@ -11,6 +11,8 @@
     autoscaling_ruler_querier_min_replicas: error 'you must set autoscaling_ruler_querier_min_replicas in the _config',
     autoscaling_ruler_querier_max_replicas: error 'you must set autoscaling_ruler_querier_max_replicas in the _config',
     autoscaling_ruler_querier_cpu_target_utilization: 1,
+    autoscaling_ruler_querier_memory_target_utilization: 1,
+    autoscaling_ruler_querier_use_oom_trigger: false,
```
Comment (jhesketh) [inline operations/mimir/autoscaling.libsonnet:15]: Personally I don't think we need to have this behind a flag unless we consider it an experimental feature.

I would rather confirm that the OOM protection works on the queries and then roll it out to all of the autoscalers.

**Review by jhesketh — COMMENTED** [review]
Just my 2cents. The rest lgtm :-)

File: operations/mimir/autoscaling.libsonnet (hunk 1 above)
Comment (chencs) [reply operations/mimir/autoscaling.libsonnet:15]: Sounds good to me! I'll remove the flag and just change the query for all autoscalers. I'm also going to change this PR back to a draft until we have good data on using the new query for the rest of the autoscalers.

File: CHANGELOG.md [hunk 2]
```diff
@@ -189,6 +190,7 @@
 * [ENHANCEMENT] Update rollout-operator to `v0.7.0`. #5718
 * [ENHANCEMENT] Increase the default rollout speed for store-gateway when lazy loading is disabled. #5823
 * [BUGFIX] Fix compilation when index, chunks or metadata caches are disabled. #5710
+* [ENHANCEMENT] Autoscaling: treat OOMing containers as though they are using their full memory request. #5739
```
Comment (jhesketh) [inline CHANGELOG.md:193]: Personally I consider this a bugfix. (But if it's not it needs moving with the other enhancements).

We can also add an enhancement here for scaling the ruler-qurier by memory.

**Review by jhesketh — CHANGES_REQUESTED** [review]
Few tweaks to the changelog, but otherwise lgtm.

We should perhaps do some deduplication in autoscaling.libsonnet of the various new*Object's, but I think that can be a followup.

**Comment by chencs** [comment]
@jhesketh dismissing your review since the changes you requested have been addressed!

**Review by 56quarters — APPROVED** [review]

**Pull request merged**